### DISCUSSION TITLE
Accessing fitted underlying scikit-learn models after fitting PyOD models

### DISCUSSION THREAD

**Discussion by AhmetZamanis** [yzhao062/pyod#512] [Q&A] [answered]
Hi, I've been experimenting with some of the PyOD anomaly detection algorithms. I have a question regarding some PyOD models which are based on scikit-learn implementations. In general, I wonder if there is a way to access the fitted underlying scikit-learn models and retrieve some attributes not available through the PyOD wrapper. I'll share two examples:

The Gaussian Mixture Model `gmm.GMM` is a wrapper around scikit-learn's `GaussianMixture`. According to the documentation, `gmm.GMM` predicts anomaly scores & binary anomaly labels, and stores various other attributes. But it does not predict the component labels as `GaussianMixture` does, or store the labels for the training data as an attribute. I wonder if it's possible to somehow retrieve the component labels so they can be easily used for "cluster plots". In comparison, it's possible to do this with `cblof.CBLOF`, through the `cluster_labels_` attribute.

The second one is the PCA-based outlier detector, `pca.PCA`. Again, this method predicts anomaly scores & labels, and stores some PCA attributes like PC loadings and variances explained. But it's not possible to retrieve the actual principal component values for the datapoints. In contrast, the scikit-learn implementation of `PCA` transforms the data into principal component values & returns these values with `.transform()`. I wonder if there is a way to retrieve the PC values either directly, or by accessing the fitted underlying `PCA` object. These could be useful to plot against the binary anomaly labels, to see how the PC values differ between anomalies and non-anomalies.

Thanks in advance for any answers.

**Comment by yzhao062** [comment, marked as answer]
PCA is also a wrapper -> https://github.com/yzhao062/pyod/blob/66010dd4cbf614b51d34bab9827151b409ae7595/pyod/models/pca.py#L230C12-L230C12 

and the attributes are available at:

![Uploading image.png…]()

**Reply by AhmetZamanis** [reply]
Thank you, I did not realize `.detector_` provides what I needed.
